### ISSUE TITLE
USB MIDI in

### ISSUE THREAD

**Issue by probonopd** [AnotherJohnH/picoX7#4]
It would be nice to have a USB MIDI in so that we can attach USB MIDI controllers (keyboards) that do not have a DIN-style MIDI connector. For this, the Pico would need to run in USB host mode.

**Comment by AnotherJohnH** [comment]
A blocker for this is I use my own SDK "Platform" that does not have USB host support. 

Two solutions...

1. Port picoX7 to another raspberry Pi Pico SDK that already has USB host support. All the hardware dependencies are deliberately kept to just [picoX7.cpp](https://github.com/AnotherJohnH/picoX7/blob/main/Source/picoX7.cpp) which should help.
2. Add USB host support to the "Platform" SDK.

TBH I'm not going to do 1. but do have a USB MIDI controller that is nicer than my actual MIDI controller. I'll have a look at the rp2040 dataset to see what's involved in adding support for USB host mode.

**Comment by AnotherJohnH** [comment]
... ah yes the world of USB has not simplified itself since I last visited many years ago!

But be e a good challenge and would open up some interesting possibilities for other RPi-Pico based projects.

**Comment by AnotherJohnH** [comment]
Currently attempting to add USB device support as a precursor to adding USB host support. If successful this should provide an easy way to control the picoX7 as a MIDI slave from a DAW. But as you state USB host support would allow direct connection to a USB-MIDI controller.

USB is complicated and MIDI-over-USB a whole other complication on top of that, so this may take a while and I might run out of enthusiasm.

However currently have the picoX7 appearing as a low level USB device, which is a good start.

**Comment by mwedmark** [comment]
Shortcut, steal with pride:
https://github.com/diyelectromusic/picodexed/tree/main/src

**Comment by AnotherJohnH** [comment]
Some small progress, picoX7 now seems to work to some degree as a proper MIDI IN device. [as of patch](https://github.com/AnotherJohnH/picoX7/commit/016d613309dc3bbf84aea4c351807efaf557bc20)
Code is provisional and needs refinement. Timing seems "lumpy"

**Comment by probonopd** [comment]
So you attach a MIDI controller (keyboard) using a USB OTG adapter cable to the USB jack on the RPi Pico, and power the RPi Pico via pins GND and VBUS?

**Comment by AnotherJohnH** [comment]
Sorry (again) should have been clearer with my "seems to work" comment. MIDI in works but only as a USB device not as a USB host. i.e. MIDI out from my laptop over USB to the picoX7 works.

**Comment by probonopd** [comment]
Thanks for the clarification. Do you plan to add USB host mode as well eventually?

**Comment by AnotherJohnH** [comment]
Yep, hope to, but my current priority is to add support for the Pico-2 and the Pimoroni pico Audio.

**Comment by mwedmark** [comment]
Sound great! I just bought a Pico2 and have the bigger brother picodemo so that combination would be cool to try. I guess that extra processing power will get many more tones poly. Maybe even support with 2 instances would be possible? ;-)

**Comment by probonopd** [comment]
Possibly https://github.com/rppicomidi/midi2usbhost/tree/main/C-Code could be helpful? It is C code that does what we need.

**Comment by probonopd** [comment]
@AnotherJohnH what do you think, could it work?

**Comment by AnotherJohnH** [comment]
Possibly, however I have a personal preference of not including any significant "third party" code in my projects. The RPiPico boot ROM and a few gcc intrinsics being the current exceptions in this project. Feel free to branch and try to integrate that code yourself.

TBH I'm finding physical MIDI in and USB device MIDI in from a DAW satisfy my needs so my desire to spend the time required adding USB host support is still not a priority and would rather spend my time to improve the sound engine simulation.
